nixos/modules/services/networking/6tunnel.nix

```
@@ -0,0 +1,111 @@
+{ lib, config, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services._6tunnel;
+
+in {
+  options.services._6tunnel = {
+
+    tunnels = mkOption {
+      default = [];
+      description = "Which tunnels to create.";
+      type = with types; listOf (submodule {
+        options = {
+          port = mkOption {
+            description = "Local port to open.";
+            type = types.port;
+          };
+          host = mkOption {
+            description = "Remote host to tunnel to.";
+            type = types.str;
+          };
+          remotePort = mkOption {
+            default = null;
+            description = "Port on remote host. If null (default), it uses the the local port.";
+            type = with types; nullOr port;
+          };
+          restart = mkOption {
+            default = null;
+            description = ''
+              Whether to restart the tunnel periodically, which is useful for non-static IP addresses.
+              The format is described in SYSTEMD.TIME(7).
+
+              If null (default), the tunnel won't restart periodically.
+            '';
+            type = with types; nullOr str;
+          };
+        };
+      });
+      example = [
+        { port = "8080"; host = "192.168.0.2"; remotePort = 80; }
+        { port = "443"; host = "www.example.com"; }
+        { port = "8053"; host = "dynamicip.example.com"; restart = "1h"; }
+      ];
+    };
+
+    openFirewall = mkOption {
+      default = true;
+      description = "Whether to open the needed firewall ports.";
+      type = types.bool;
+    };
+
+  };
+
+  config = mkIf (cfg.tunnels != {}) {
+
+    networking.firewall.allowedTCPPorts = mkIf (cfg.openFirewall) (map (tunnel: tunnel.port) cfg.tunnels);
+    networking.firewall.allowedUDPPorts = mkIf (cfg.openFirewall) (map (tunnel: tunnel.port) cfg.tunnels);
+
+    systemd.services = listToAttrs (map (tunnel:
+      nameValuePair "6tunnel-${toString tunnel.port}" {
+        description = "6tunnel from ${toString tunnel.port} to ${tunnel.host}${if (tunnel.remotePort != null) then ":${toString tunnel.remotePort}" else ""}";
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        serviceConfig = {
+          ExecStart = ''
+            ${pkgs._6tunnel}/bin/6tunnel -d ${toString tunnel.port} "${tunnel.host}" ${optionalString (tunnel.remotePort != null) (toString tunnel.remotePort)}
+          '';
+          Restart = "always";
+          RuntimeMaxSec = mkIf (tunnel.restart != null) tunnel.restart;
+          # Hardening
+          DynamicUser = true;
+          Group = "nogroup";
+          AmbientCapabilities = "CAP_NET_BIND_SERVICE";
```

```suggestion
          AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
```
Are you sure this works with `CapabilitiesBindingSet = ""`? A test would be nice.

---

from https://lists.freedesktop.org/archives/systemd-devel/2018-September/041386.html:
> They're opposites. AmbientCapabilities= will *grant* the listed
capabilities automatically, while CapabilityBoundingSet= will *limit* what
capabilities the process may receive from any sources (even limiting suid).

I've tested multiple ports and everything seems to work.

---

Still, I think we should write a very simple test, e.g. using https://docs.python.org/3/library/http.server.html.

---

Will be my first test, I'll give it a shot.

---

Thanks! If it doesn't work out, I can also write it. Shouldn't be much work.

---

I had something else to do in the meantime, so it took a little longer, but I'm finished now. Can you have a look please?

---

Looks good! Thanks a lot.